doc/manual/source/user_guide/InSituPythonAnalysis.rst

```diff
@@ -0,0 +1,269 @@
+
+How it Works
+------------
+Enzo follows ``libyt``'s procedure and APIs to implement this in situ analysis feature.
+We can find how ``libyt`` works more in detail `here <https://libyt.readthedocs.io/en/latest/how-it-works.html>`__ and what are ``libyt`` APIs `here <https://libyt.readthedocs.io/en/latest/libyt-api/index.html>`__.
+
+At initialization stage, ``libyt`` imports inline Python script ``inline.py`` and initializes Python interpreters in each MPI process. This happens in ``InitializeLibytInterface`` function in ``src/enzo/InitializeLibytInterface.C``.
+
+When Enzo finishes its computation in a cycle, the whole simulation pauses and starts the in situ analysis process.
+The ``CallInSitulibyt`` function in ``src/enzo/CallInSitulibyt.C`` conducts this process.
+Enzo then passes in simulation information and actual field data pointers to ``libyt``.
+This includes simulation information, like adaptive mesh grid hierarchy, parameters, field labels, etc, and actual simulation data pointers inside ``BaryonField`` array.  (Data fields are referenced, not copied, from ``libyt``'s perspective, although typically ``yt`` itself will make a copy as needed before any in-place changes occur.)
+``libyt`` will construct data structures to store simulation information and wrap these data pointers, so that they can be read and used in Python with minimum memory overhead.
+
+``libyt`` supports calling Python functions from simulation process,
+and it also supports user interface (**interactive mode** has :ref:`Interactive Python Prompt` and :ref:`Reloading Script` feature, and **jupyter kernel mode** has :ref:`Jupyter Notebook / JupyterLab UI` feature).
+We can update Python functions and probe data interactively in the UI.
+After in situ analysis is done, ``libyt`` frees resources allocated for itself, and the simulation will continue.
+
+At the end, when the simulation is shutting down, Enzo calls ``FinalizeLibytInterface`` in ``src/enzo/InitializeLibytInterface.C``, so that ``libyt`` finalizes the Python interpreters.
+
+How to Configure
+----------------
+**Some settings are hard-coded inside Enzo, we can customize it to our own needs.**
+
+General
+^^^^^^^
+
+* **How to change import Python file name?**
+
+  The default Python script will be imported is ``inline.py``.
+  If we really want to change the name, we can go to
+  ``src/enzo/InitializeLibytInterface.C`` in function ``InitializeLibytByItself``, and change ``params->script`` to our Python file name without ``.py``. For example, I want to make it to ``test.py``:
+
+  ::
+
+      params->script = "test";
+
+  Please use ``const char*``, or else, we have to make sure the lifetime of this variable covers the whole in situ process.
+
+* **How to activate in situ Python analysis process?**
+
+  The full process is encapsulated inside ``CallInSitulibyt`` function in ``src/enzo/CallInSitulibyt.C``.
+  We can put this function everywhere we want in Enzo to start in situ analysis.
```

Comment:
I suspect that `CallInSitulibyt` cannot be called everywhere, for example it needs to have `LevelArray` accessible.  Probably this can only be called within `EvolveLevel`?

Reply:
@matthewturk, Can you help with this?

Reply:
The existing uses of `CallInSitulibyt` are indeed only in `EvolveLevel` and after the recursive loop. Given that the simulation is supposed to finish evolution before handing over data, this makes sense. I don't think `CallInSitulibyt` should (much less can) be added to any other location because of this as it would be unwise to allow data to be passed to in situ analysis while the level hierarchy is still being evolved. Though, see my notes on `EvolveLevel.C` for my thoughts on its current placement.

Overall, I don't think the docs should advertise that `CallInSitulibyt` can be added anywhere else than where it already is.

Reply:
Whoops, this comment was supposed to say "``CallInSitulibyt`` **can't** be added anywhere else than where it already is.